### PR TITLE
Feature/delayed bcf export

### PR DESCRIPTION
**Pull request by ulferts**
Instead of attempting to export a set of work packages to the bcf xml format synchronously, with the browser waiting for the response, the browser now:

* issues a triggering request
* polls for the response to be finished
* downloads the export 

During that time, the user is presented with a waiting screen:
image

The export file is stored as an attachment, attached to a newly introduced class `WorkPackages::Export`. The attachment is only accessible by the user triggering the export as the export contains work package data which is different for every user. The export is removed automatically after a grace period (same configuration as for uncontainered attachments `attachments_grace_period`). 

Exports to other formats (e.g. pdf) are still created synchronously although those might also benefit from being implemented asynchronously. But because the front end does not know whether it will receive a delayed or synchronous export, it has to first trigger the export with one request and then download it in the next request (done automatically for the user). To avoid exporting twice, all exports (i.e. the synchronous ones like PDF as well) are now stored as attachments. One could have implemented this differently but this structure should ease transitioning to a delayed export for all exports later on.

https://community.openproject.com/projects/openproject/work_packages/33040

### CODE REVIEW

File: frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts [hunk 1]
```diff
@@ -1,23 +1,40 @@
...
+The backend may choose to provide the export right away (synchronously) or delayed. In the later case, the modal will poll the
+backend until the export is done.
+Because the modal has to deal with both cases, without knowing upfront whether the download will be delayed or not, it needs
+to treat every download expecting it to be delayed. Because of this, the modal cannot simply provide download hrefs which would
+ allow the browser to download the export but rather has to first check whether an export is delayed or not, and if it is delayed,
+ it has to wait until the export is ready. Because of the necessary check, the modal has a hidden link that it clicks on to perform
+ the actual download once the export is ready (delayed or not).
+
+The modal might also be used to only display the progress of an export. This will happen if a link for exporting is provided via the locals.
+ */
```
Comment (openprojectci) [inline frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts:32]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/5cc0b80f0d93730810e407445c446a54cf5a577c/frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts#L32"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

File: frontend/src/app/modules/bim/ifc_models/toolbar/import-export-bcf/bcf-export-button.component.ts [hunk 2]
```diff
@@ -35,13 +35,15 @@ import {QueryResource} from "core-app/modules/hal/resources/query-resource";
 import {UrlParamsHelperService} from "core-components/wp-query/url-params-helper";
 import {StateService} from "@uirouter/core";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {WpTableExportModal} from "core-components/modals/export-modal/wp-table-export.modal";
+import {OpModalService} from "core-components/op-modals/op-modal.service";
```
Comment (openprojectci) [inline frontend/src/app/modules/bim/ifc_models/toolbar/import-export-bcf/bcf-export-button.component.ts:40]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/97c28f01d11b11486bd4209ccf1016cee9e89fb5/frontend/src/app/modules/bim/ifc_models/toolbar/import-export-bcf/bcf-export-button.component.ts#L40"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

File: frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts [hunk 3]
```diff
@@ -63,6 +102,47 @@ export class WpTableExportModal extends OpModalComponent implements OnInit {
...
+    let poller = this
+      .httpClient
+      .get(url, { observe: 'response'})
+      .pipe(
+        catchError(this.errorOrDownload.bind(this))
+      );
+
+    this.subscription = poller.pipe(
+      expand(_ => timer(1000).pipe(concatMap(_ => poller)))
+    ).subscribe();
```
Comment (ulferts) [inline frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts]: @oliverguenther there is still the console error at the end of a download:

image

As your Angular foo is stronger, maybe you can figure out the problem.

**Review by oliverguenther — APPROVED** [review]
Looks good to me. I still need to look at the console error, but will do so in a separate PR if necessary

File: frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts [hunk 4]
```diff
@@ -1,23 +1,44 @@
...
+The backend may choose to provide the export right away (synchronously) or delayed. In the later case, the modal will poll the
+backend until the export is done.
+Because the modal has to deal with both cases, without knowing upfront whether the download will be delayed or not, it needs
+to treat every download expecting it to be delayed. Because of this, the modal cannot simply provide download hrefs which would
+ allow the browser to download the export but rather has to first check whether an export is delayed or not, and if it is delayed,
+ it has to wait until the export is ready. Because of the necessary check, the modal has a hidden link that it clicks on to perform
+ the actual download once the export is ready (delayed or not).
+
+The modal might also be used to only display the progress of an export. This will happen if a link for exporting is provided via the locals.
+ */
```
Comment (openprojectci) [inline frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts:36]: <table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/97c28f01d11b11486bd4209ccf1016cee9e89fb5/frontend/src/app/components/modals/export-modal/wp-table-export.modal.ts#L36"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

**Comment by ulferts** [comment]
Because it was actually easier in the end, I turned all work package list exports to be delayed now. We can now think about removing the export limit setting for the file based exports but need to bear in mind that we still use it in the API.

**Pull request merged**